### DISCUSSION TITLE
Selfhost Setup

### DISCUSSION THREAD

**Discussion by ABouveron** [keygen-sh/community#87] [Q&A] [answered]
Hello,

After having struggled a bit on the selfhost setup of the Community Edition, I thought I was finally done but I can't seem to communicate with api.keygen.selfhost, with the same queries that I use with the official API. 
The example that I mostly tried to use is this one: https://keygen.sh/docs/api/tokens/#tokens-generate 

I'm always receiving this error though: 

> {"meta":{"id":"b533808a-8ae5-4ac6-98c6-eb87ba498bc2"},"errors":[{"title":"Not found","detail":"The requested endpoint was not found (check your HTTP method and resource path)","code":"NOT_FOUND"}]}

The same request work with api.keygen.sh and I believe Caddy is working correctly despite the two following warnings: `admin endpoint disabled` and `stapling OCSP   {"error": "no OCSP stapling for [api.keygen.localhost]: no OCSP server specified in certificate", "identifiers": ["api.keygen.localhost"]}`. 

Web.1 logs:

> method=POST path=/v1/accounts/d9a5119f-f9ca-44fd-8920-a5f9e21f7258/tokens format=jsonapi controller=ErrorsController action=show status=404 duration=3.87 view=0.11 host=api.keygen.localhost request_id=b533808a-8ae5-4ac6-98c6-eb87ba498bc2 api_version=1.3 query_params=N/A account_id=N/A account_slug=N/A env_id=N/A env_code=N/A bearer_type=N/A bearer_id=N/A token_id=N/A authn=N/A authz=N/A ip=127.0.0.1 user_agent=RestSharp/110.2.0.0 origin=N/A time=2023-07-13 12:12:00 UTC code=NOT_FOUND encoded_response=eyJtZXRhIjp7ImlkIjoiYjUzMzgwOGEtOGFlNS00YWM2LTk4YzYtZWI4N2JhNDk4YmMyIn0sImVycm9ycyI6W3sidGl0bGUiOiJOb3QgZm91bmQiLCJkZXRhaWwiOiJUaGUgcmVxdWVzdGVkIGVuZHBvaW50IHdhcyBub3QgZm91bmQgKGNoZWNrIHlvdXIgSFRUUCBtZXRob2QgYW5kIHJlc291cmNlIHBhdGgpIiwiY29kZSI6Ik5PVF9GT1VORCJ9XX0= rate_limited=false rate_reset=2023-07-13 12:12:30 +0000 rate_window=30 rate_count=1 rate_limit=60 rate_remaining=59

I don't see anything special here besides the 404 error.

I can still ping the API though: 

> method=GET path=/v1/ping format=html controller=Api::V1::HealthController action=general_ping status=200 duration=34.61 view=0.20 db=10.71 host=api.keygen.localhost request_id=d44a287b-54c6-4645-872c-f8f0f902e78e api_version=1.3 query_params=N/A account_id=N/A account_slug=N/A env_id=N/A env_code=N/A bearer_type=N/A bearer_id=N/A token_id=N/A authn=N/A authz=N/A ip=127.0.0.1 user_agent=Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/115.0 origin=N/A time=2023-07-13 12:14:04 UTC code=N/A encoded_response=N/A rate_limited=false rate_reset=2023-07-13 12:14:30 +0000 rate_window=30 rate_count=1 rate_limit=60 rate_remaining=59

Am I missing something in the Self Hosting doc, or expecting too much (do we need to code our own endpoints?)?
Any help would be appreciated, I can give more information if that would help you.

**Comment by ezekg** [comment, marked as answer]
Is `KEYGEN_HOST` set to `api.keygen.selfhost`? Without that, the `/v1/accounts/<account>/*` namespace can't route correctly. You can likely still use `/v1/tokens` (without the account namespace). But I'd confirm that your environment variables are set up correctly first.

**Comment by ABouveron** [comment]
Oh I wanted to say `api.key.localhost`, sorry about that. But yeah it was the problem, I followed the doc blindly and since I didn't define `KEYGEN_HOST` on my end, the line `KEYGEN_HOST="${KEYGEN_HOST}"` didn't do much lol. Thank you for the help!

By the way, I saw on several discussions that you were planning to release a selfhosted dashboard in the near future... Do you have any idea when it would be available?

**Reply by ezekg** [reply]
No ETA yet, sorry. But actively working on it.
